tests/test_responsetypes.py

```diff
@@ -72,17 +72,33 @@ def test_from_headers(self):
             retcls = responsetypes.from_headers(source)
             assert retcls is cls, f"{source} ==> {retcls} != {cls}"
 
-    def test_from_args(self):
-        # TODO: add more tests that check precedence between the different arguments
+    def test_from_args_pre_xtractmime(self):
         mappings = [
             ({'url': 'http://www.example.com/data.csv'}, TextResponse),
-            # headers takes precedence over url
             ({'headers': Headers({'Content-Type': ['text/html; charset=utf-8']}),
               'url': 'http://www.example.com/item/'}, HtmlResponse),
             ({'headers': Headers({'Content-Disposition': ['attachment; filename="data.xml.gz"']}),
               'url': 'http://www.example.com/page/'}, Response),
+        ]
+        for source, cls in mappings:
+            retcls = responsetypes.from_args(**source)
+            assert retcls is cls, f"{source} ==> {retcls} != {cls}"
 
-
+    def test_from_args_post_xtractmime(self):
+        mappings = [
+            ({'body': b'Some plain\0 text data with\0 tabs and null bytes\0'}, TextResponse),
+            ({'body': b'\x03\x02\xdf\xdd\x23', 'headers': Headers({'Content-Encoding': 'UTF-8'})},
+             Response),
+            # different behaviour with http and non-http urls
+            ({'body': b'\x00\x01\xff', 'url': 'http://www.example.com/item/',
+              'headers': Headers({'Content-Type': b'text/plain'})}, Response),
+            ({'body': b'\x00\x01\xff', 'url': '://www.example.com/item/',
+              'headers': Headers({'Content-Type': b'text/plain'})}, TextResponse),
+            ({'url': 'http://www.example.com/item/file.html'}, HtmlResponse),
+            ({'body': b'Some plain text data\1 with tabs and\n null bytes\0'}, Response),
+            ({'body': b'<html><head><title>Hello</title></head>'}, HtmlResponse),
+            ({'body': b'<?xml version="1.0" encoding="utf-8"'}, XmlResponse),
+            ({'filename': 'file.html'}, HtmlResponse),
```

Comment:
I think some of these tests would have also passed with the pre-xtractmime implementation, in which case you should move them to `test_from_args_pre_xtractmime`; let’s keep here only the expected behavioral changes.